tests/run-pass/concurrency/tls_lib_drop.rs

```diff
@@ -0,0 +1,46 @@
+// ignore-windows: Concurrency on Windows is not supported yet.
+
+//! Check that destructors of the library thread locals are executed immediately
+//! after a thread terminates.
+
+#![feature(thread_local_internals)]
+
+use std::cell::RefCell;
+use std::thread;
+
+struct TestCell {
+    value: RefCell<u8>,
+}
+
+impl Drop for TestCell {
+    fn drop(&mut self) {
+        println!("Dropping: {}", self.value.borrow())
+    }
+}
+
+static A: std::thread::LocalKey<TestCell> = {
+    #[inline]
+    fn __init() -> TestCell {
+        TestCell { value: RefCell::new(0) }
+    }
+
+    unsafe fn __getit() -> Option<&'static TestCell> {
+        static __KEY: std::thread::__OsLocalKeyInner<TestCell> =
+            std::thread::__OsLocalKeyInner::new();
+        __KEY.get(__init)
+    }
```

Comment:
Woah, that's some internal API usage.^^ Is this not exposed?

Also, should there be a similar test for tls_lang? Does that even have destrcutors? Like, do the `Drop` impls of thread-local statics get run at some point?

Reply:
> Woah, that's some internal API usage.^^ Is this not exposed?

At least I have not found where it would be properly exposed. This is what `thread_local!` macro emits on some platforms.

> Also, should there be a similar test for tls_lang? Does that even have destrcutors? Like, do the `Drop` impls of thread-local statics get run at some point?

This example:
```rust
#[thread_local]
static X: u8 = 0;
```
gives the following error:
```
error[E0658]: `#[thread_local]` is an experimental feature, and does not currently handle destructors
```
Therefore, I think that the thread-local statics do not support destructors yet. (I have already added a `FIXME` pointing out that destructors for the thread-local statics are currently missing.)

Reply:
> This is what thread_local! macro emits on some platforms.

On *some* platforms? But we cannot use it here because it doesn't do that on Linux?

> Therefore, I think that the thread-local statics do not support destructors yet.

That makes me very happy actually. :D  So dtors are limited to library-TLS, and not present in language-TLS. Good.

Reply:
> On some platforms?

Yes. It is chosen for platforms on which thread-local statics do not have destructors.

> But we cannot use it here because it doesn't do that on Linux?

Actually, I misread the documentation. We can use it on Linux. I have updated the test.

Reply:
> Yes. It is chosen for platforms on which thread-local statics do not have destructors.

Wait, I thought the conclusion was that thread-local statics *do not* have destructors?
If they do on some platforms, doesn't that mean we need to implement that (and find some way to test it)?

Reply:
Oh okay so looks like the `thread_local!` macro, when possible, will use a thread-local static for the *data* but use library TLS stuff for the *destructor*? Wow, I hope we have covered all the cases now...

Reply:
> Wait, I thought the conclusion was that thread-local statics do not have destructors?

As far as I understand, thread-local statics do not have destructors **yet** (at least I could not make them run on the playground). Once they do, we will have to implement them. However, the test based on the `thread_local!` macro is likely to start failing at that point, so we should not miss that moment.

> Oh okay so looks like the thread_local! macro, when possible, will use a thread-local static for the data but use library TLS stuff for the destructor?

At least that is also my current understanding.

> Wow, I hope we have covered all the cases now...

I hope so too...